Add wrapper to `kn` CLI to allow potentially downloading newer version from cluster

amisevsk opened the pull request:
### Description
This PR adds a wrapper, `kn`, which is placed earlier in the path than the already-installed `kn` CLI. When the user first runs `kn`, this wrapper runs instead and, if the cluster has the serverless operator installed, prompts the user to download the `kn` CLI served by the operator instead of using the local version. If the user agrees to download the CLI, it is downloaded and run instead of the installed version of `kn`.

If the user responds to the prompt with "yes" or "no" (i.e. doesn't leave it blank), this preference is saved to a file inside the container (for speed of access) and to an annotation `web-terminal.redhat.io/kn-wrapper-preference` (to persist the decision between terminal starts).

This PR also changes the entrypoint.sh of the container to call this wrapper and automatically re-download `kn` if that is the preference the user chose.

The goal of the wrapper is to be transparent to the user except when prompting is required. As a result, we take the following behavior:
* If the serverless operator is not installed on the cluster, or the current user can't read the download URL for any reason, we silently pass arguments to the wrapper to the already-installed `kn` CLI
* If the user responds n/N/no/No to the prompt, we save this preference and only ever call the already-installed CLI with arguments.
* If the user responds y/Y/yes/Yes to the prompt, we download (and re-download when the terminal restarts) the kn CLI and transparently call that in the future.

The wrapper hides three wrapper-specific arguments to make debugging easier:
```
  --wto-reset-preferences   - Reset any saved preferences and remove existing
                              downloaded 'kn' binary, if it exists.
  --wto-restore-preferences - Non-interactively read preferences and download
                              'kn' from the cluster if necessary. This is used
                              for initial setup when the Web Terminal starts.
  --wto-show-logs           - Print internal log file to standard out
```

### How to test
Changes from this PR as built as image `quay.io/amisevsk/web-terminal-tooling:kn-wrapper`. To test these changes, execute the following command in any running web terminal:
```bash
wtoctl set image quay.io/amisevsk/web-terminal-tooling:kn-wrapper
```
(preferably in a cluster with serverless installed, e.g. developer sandbox)

Then run the `kn` CLI in the terminal once it restarts.

### Gif
![kn-wrapper](https://github.com/redhat-developer/web-terminal-tooling/assets/16168279/20744ad9-f32c-4c00-93d9-9c180336add2)

cardil commented:
I got a failure caused by untrusted TLS cert of a cluster. Looks like the WTO terminal lacks the certs that are installed on the cluster.

```
ash-4.4 ~ $ kn version
Detected serverless installation in this cluster.
The 'kn' CLI is available at: 
Would you like to automatically download 'kn' from this URL instead of using the built-in version? (y/N): y

Failed to download CLI. See log file /wto/bin/kn-wrapper.log for details
bash-4.4 ~ $ cat /wto/bin/kn-wrapper.log                         
[2023-10-30T08:26:13Z INFO ] No preference file found, attempting to read preference from DevWorkspace annotations
[2023-10-30T08:26:14Z INFO ] Could not find preference in DevWorkspace annotations
[2023-10-30T08:26:14Z INFO ] No preference file found, attempting to read preference from DevWorkspace annotations
[2023-10-30T08:26:15Z INFO ] Could not find preference in DevWorkspace annotations
[2023-10-30T08:26:15Z INFO ] Found consoleclidownload for kn with URL https://kn-openshift-serverless.apps.jeramos-590.ci-psi.rhocf-dev.net/kn-linux-amd64.tar.gz
[2023-10-30T08:26:37Z INFO ] No preference file found, attempting to read preference from DevWorkspace annotations
[2023-10-30T08:26:38Z INFO ] Could not find preference in DevWorkspace annotations
[2023-10-30T08:26:38Z INFO ] Found consoleclidownload for kn with URL https://kn-openshift-serverless.apps.jeramos-590.ci-psi.rhocf-dev.net/kn-linux-amd64.tar.gz
[2023-10-30T08:26:42Z INFO ] Received response 'y' for prompt
[2023-10-30T08:26:42Z INFO ] Downloading kn from https://kn-openshift-serverless.apps.jeramos-590.ci-psi.rhocf-dev.net/kn-linux-amd64.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: https://curl.haxx.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
```

l0rd commented:
@cardil can you test if setting `export CURL_CA_BUNDLE = "/etc/ssl/certs/ca-bundle.trust.crt"` or ``export CURL_CA_BUNDLE = "/public-certs/kube-root-ca.crt.ca.crt"` makes `curl https://kn-openshift-serverless.apps.jeramos-590.ci-psi.rhocf-dev.net/kn-linux-amd64.tar.gz` work?

amisevsk commented:
I've updated the PR with the following changes:

* After some discussion with @l0rd (thanks for the suggestion!), the wrapper will now use the mounted certificate at `/var/run/secrets/kubernetes.io/serviceaccount/ca.crt` when attempting to download `kn` -- this should resolve the issue you faced @cardil. I've tested this on a cluster with self-signed certificates and the wrapper was able to download `kn` without issue
* I've reworked how setup/preferences works so that we also store the URL we previously used to download `kn`. If this URL changes, the user is prompted to accept the new URL. This is done to make sure we only download `kn` from URLs the user explicitly trusts.
* I also added more logging (to the file) and made sure the wrapped `kn` still works in e.g. scripts, to avoid issues where e.g. a non-interactive script hangs indefinitely waiting for input that will not come.

Please take a look when you have time.

cardil commented:
@amisevsk Works well, thanks! Just tested that on not-yet released kn, and it worked.

ibuziuk reviewed (approved):
verified on Developer Sandbox
